src/ray/util/event.cc

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/event.h"
+
+namespace ray {
+thread_local std::unique_ptr<RayEventContext> RayEventContext::instance_ = nullptr;
+
+LogBasedEventReporter::LogBasedEventReporter(rpc::Event_SourceType source_type, std::string log_dir)
+    : log_dir_(log_dir) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
+  file_name_ = "event_" + Event_SourceType_Name(source_type);
+  if (source_type == rpc::Event_SourceType::Event_SourceType_CORE_WORKER ||
+      source_type == rpc::Event_SourceType::Event_SourceType_COMMON) {
+    file_name_ += "_" + std::to_string(getpid());
+  }
+}
+
+std::string LogBasedEventReporter::EventToString(rpc::Event &event) {
+  std::stringstream result;
+
+  struct tm *dt;
+  char time_buffer[30];
+  time_t epoch_time_as_time_t = event.timestamp() / 1000;
+
+  dt = localtime(&epoch_time_as_time_t);
+  strftime(time_buffer, sizeof(time_buffer), "%Y-%m-%d %H:%M:%S.", dt);
+
+  result << std::string(time_buffer) << std::setw(5) << std::setfill('0')
+         << event.timestamp() % 1000 << separator_ << event.event_id() + separator_
+         << event.job_id() + separator_ << event.node_id() + separator_
+         << event.task_id() + separator_
+         << Event_SourceType_Name(event.source_type()) + separator_
+         << event.source_hostname() + separator_
+         << std::to_string(event.source_pid()) + separator_
+         << Event_Severity_Name(event.severity()) + separator_
+         << event.label() + separator_ << event.message();
+  return result.str();
+}
+
+void RayEventContext::SetEventContext(rpc::Event_SourceType source_type,
```

Comment:
cc @edoakes I think it is better having more flexible map labels rather than 4 hardcoded labels. The idea is we always have global level labels per process (like these 4), and support additional global level labels or custom labels at each event.

Reply:
we prepare to add a map<string, string> to store the custom labels 
maybe the user can use RayEventContext::Instance()::SetCustomContext(std::string, std::string) 
I will modify this week.

Reply:
Yes I would much prefer to just have a map of labels. We can autopopulate a few of them but I don't see a reason to have these hardcoded.